### ISSUE TITLE
Property vs. DynamicSubject/Bond

### ISSUE THREAD

**Issue by connectdotz** [ReactiveKit/Bond#291]
Hello,

As part of our swift3 migration, we are transitioning from ReactiveUIKit to Bond and noticed some seemingly by-design difference, just trying to understand the rationale so we can adjust our logic accordingly:

in ReactiveUIKit, most reactive UI attributes are implemented as <code>Property</code>, which means they are both signal producer and consumer with its own getter/setter. We have used this characteristics to implement something like in <code>UIButton</code>, when <code>rEnabled</code> is changed, we will also change the button's background color and title accordingly by observing the property.

However in Bond, all the reactive UI attributes are either <code>DynamicSubject</code> (which is both producer/consumer but without getter/setter), or <code>Bond</code> (just a consumer which can't be observed directly). The UIButton's <code>bnd_isEnabled</code> is unfortunately a Bond attribute so we can't observe it to chain other changes like we did before... a few other issues:
1. It wasn't very clear as which attributes are <code>DynamicSubject</code> or  <code>Bond</code>, especially they all have the same naming convention; so we either have to look up on the source code or settle for trial-and-error, which is obviously bad.
2. we wonder if it will be more intuitive to create higher level attributes, like the <code>Property</code> in ReactiveUIKit, which will eliminate the "role" confusion and support more use cases right out of boxes. Maybe there are downsides we were not aware of, if you don't mind explaining that will be really helpful.

**Comment by jechol** [comment]
I feel same. I don't understand why many of `bnd_xxx` became just consumer.

Hard to migrate.

**Comment by srdanrasic** [comment]
Yes, this is by design. We have two types here as you have correctly inferred.

1. `DynamicSubject` - both a producer and a consumer.
2. `Bond` - just a consumer. 

Reason Bond is just a consumer is that the underlying component doesn't produce anything. UILabel cannot change its text property on its own. 

On the other hand, UITextField changes its text property when user writes/modifies text. Any component that produces values (basically any input component, i.e. a component that reads data from the user) is represented with DynamicSubject. That DynamicSubject usually wraps Target-Action or Delegate pattern.

Let's say that you have a login button and an info label that is hidden only when button is enabled. My guess is that you've been doing something like this:

```swift
viewModel.buttonEnabled.bind(to: button.bnd_enabled)
button.bnd_enabled.bind(to: textLabel.bnd_hidden)
```

Problem is that you're deriving state of one view from the state of another view. With that you're creating a chain of bindings and large chains are bad. They can be hard to debug. You change one thing, five others change to.

Instead, we should always derive state from the place where the underlying truth lives, in our case from the view model. So we can rewrite that code to something like this:

```swift
viewModel.canLogin.bind(to: button.bnd_enabled)
viewModel.canLogin.bind(to: textLabel.bnd_hidden)
```

In other words, we were thinking "If logging in is possible, enable the button. Also, if button is enabled show the label" while we should have been thinking "If logging in is possible enable the button and hide the label".

This is trivial example, but shows the preferred way to go. Just replace `X -> Y; Y -> Z` to `X -> Y; X -> Z`.  Hope it helps.

**Comment by connectdotz** [comment]
@srdanrasic, thanks for the explanation. 

I completely understand what you are saying and agree that it is indeed the right way of thinking about some use cases. However, there are some other use cases that might not fit quite as obvious. For example, when multiple ViewModels needs to trigger the same view behavior, such as enable/disable a commit button when any form field changes. One could argue the button's background and title is actually depends on the button's own state, not necessary the models affect the button, i.e. the "truth" is the button's own state. So it doesn't make much sense for the models to reach into the button's own background and title with binding...

Obviously we need to be careful with such reasoning so it won't become unnecessary spaghetti code as you are rightfully trying to avoid. The question is should the library dictate the use case or provide the flexibility and let developer decide how to use them (+ best practice to warm cases like you mentioned)

Regarding the DynamicSubject vs Property: I see one more difference that you explicitly distinguish "how" the text changed: triggered by events vs. manually setting. With <code>Property</code>, it handles both cases internally, i.e. it can be observed for changes caused by UI or by setting value manually. With <code>DynamicSubject</code>, it only emit signal for UI triggered changes. 

I can see your intend to simplify and enforce proper patterns, that's all good; and sure we can just add the additional properties for our own use cases, but if this is a common pattern (maybe?...:-)), I wonder will it be more benefitial to add it in the library?

**Comment by jechol** [comment]
@srdanrasic Totally agree now.

What makes me hard to migrate was wrong state derivation path. Thanks for the explanation.

**Comment by IceFloe** [comment]
Hi all! Just want to say, that I can agree with @connectdotz that this is a problem, now I have to write more code is some situations, this is rly painful. Let see some example 
```
Before:
        viewModel.iDontCareFilter.bidirectionalBindTo(doesNotMatterFilterBtn.bnd_selected)
        viewModel.lowFilter.bidirectionalBindTo(lowFilterBtn.bnd_selected)
        viewModel.mediumFilter.bidirectionalBindTo(mediumFilterBtn.bnd_selected)
        viewModel.hardFilter.bidirectionalBindTo(hardFilterBtn.bnd_selected)
        viewModel.insaneFilter.bidirectionalBindTo(insaneFilterBtn.bnd_selected)

@IBAction func changeFilterState(_ sender: FilterButton) {
        sender.bnd_selected.value = !sender.bnd_selected.value
    }

```
After:
```
        viewModel.iDontCareFilter.bind(to: doesNotMatterFilterBtn.bnd_isSelected)
        viewModel.lowFilter.bind(to: lowFilterBtn.bnd_isSelected)
        viewModel.mediumFilter.bind(to: mediumFilterBtn.bnd_isSelected)
        viewModel.hardFilter.bind(to: hardFilterBtn.bnd_isSelected)
        viewModel.insaneFilter.bind(to: insaneFilterBtn.bnd_isSelected)

@IBAction func changeFilterState(_ sender: FilterButton) {
        if sender == self.doesNotMatterFilterBtn {
            viewModel.iDontCareFilter.next(!sender.isEnabled)
        }
        
        if sender == self.lowFilterBtn {
            viewModel.lowFilter.next(!sender.isEnabled)
        }
        
        if sender == self.mediumFilterBtn {
            viewModel.lowFilter.next(!sender.isEnabled)
        }
        
        if sender == self.hardFilterBtn {
            viewModel.lowFilter.next(!sender.isEnabled)
        }
        
        if sender == self.insaneFilterBtn {
            viewModel.insaneFilter.next(!sender.isEnabled)
        }
}
```

**Comment by srdanrasic** [comment]
Thanks everyone for your feedback. I see the inconvenience. I'm working on some ideas.

**Comment by AnthonyMDev** [comment]
Thanks for considering addressing this issue @srdanrasic. I'd like to give some input as well if that's alright.

I agree that these bindings should be `Property` type instead of `DynamicSubjects`. Being able to manually set these properties and know that they will emit events when they are set is a valuable feature.

As for @connectdotz other point about the button state:

> One could argue the button's background and title actually depend on the button's own state, not necessary the models affect the button, i.e. the "truth" is the button's own state. So it doesn't make much sense for the models to reach into the button's own background and title with binding...

Couldn't this be solved by creating a new state or even a simple property on your view elements yourself? If you want one property that will change multiple properties on the button you can do this.

````swift
let buttonEnabledState: Property<Bool> =  {
    let property = Property(true)
    button.bnd_enabled.bind(to: property)
    button.bnd_hidden.bind(to: property.map { !$0 })
    return property
}()
````

Then you can bind your view model to the new property.

````swift
viewModel.buttonEnabled.bind(to: button.buttonEnabledState)
````

The `buttonEnabledState` property can be added either in a custom subclass of `UIButton` or using an associated value on an extension on `UIButton` itself.

Would this solve your problem @connectdotz or am I misunderstanding your point?

**Comment by connectdotz** [comment]
@AnthonyMDev, yes we are fully aware that adding our own property can address the problem, and that is exactly what we did to work around the current problem. My main point is: if many people end up creating many similar/dup properties, it might make more sense to have the library providing it instead. Imagine if the attributes are all <code>Property</code>, then we don't need to create any extra attributes, isn't that better?!

**Comment by srdanrasic** [comment]
Unfortunately, making extensions Properties would mean giving up on all the benefits of structs. Current implementation is extremely lightweight.

However, with some additions, they work almost as properties. Value getter is back. You can read it 

```swift
print(textField.bnd_text.value)
```

and if you want to set a new value and trigger an event, use `next` method

```swift
textField.bnd_text.next("test")
```

Hope this helps! Check out v5.1.
